### PR TITLE
osu-lazer: 2023.301.0 -> 2023.305.0

### PR DESCRIPTION
**Pull request by delan**
###### Description of changes

This patch updates osu-lazer and osu-lazer-bin from 2023.301.0 to [2023.305.0](https://github.com/ppy/osu/releases/tag/2023.305.0), and includes @StepBroBD’s changes adding darwin support to osu-lazer-bin (delan/nixpkgs#2).

I’ve extended update.sh to also update the bin package. I couldn’t figure out how to do it with update-source-version though, since trying to update other platforms gives me an error:

```
$ update-source-version osu-lazer-bin "$new_version" --system=aarch64-darwin
error: a 'aarch64-darwin' with features {} is required to build '/nix/store/...', but I am a 'x86-64_linux' with features {benchmark, big-parallel, kvm, nixos-test}
```

I’ve tested both on NixOS 22.05 (still no time to upgrade sorry) with the following additional patch:

```diff
diff --git a/pkgs/games/osu-lazer/default.nix b/pkgs/games/osu-lazer/default.nix
index 3a56c35fe88..ff37f988ccf 100644
--- a/pkgs/games/osu-lazer/default.nix
+++ b/pkgs/games/osu-lazer/default.nix
@@ -29,6 +29,10 @@ buildDotnetModule rec {
   projectFile = "osu.Desktop/osu.Desktop.csproj";
   nugetDeps = ./deps.nix;
 
+  # TODO remove in NixOS 22.11
+  dotnet-sdk = dotnetCorePackages.sdk_6_0;
+  dotnet-runtime = dotnetCorePackages.runtime_6_0;
+
   nativeBuildInputs = [ copyDesktopItems ];
 
   runtimeDeps = [
```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] ~~aarch64-linux~~
  - [ ] x86_64-darwin (untested!)
  - [ ] aarch64-darwin (untested!)
- [ ] ~~For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))~~
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/games/osu-lazer/update.sh [hunk 1]
```diff
@@ -15,7 +16,22 @@ fi
...
+    'x86_64-linux osu.AppImage' \
+  ; do
+    set -- $pair
+    nix-prefetch-url "https://github.com/ppy/osu/releases/download/$new_version/$2" | while read -r hash; do
+      nix --extra-experimental-features nix-command hash to-sri --type sha256 "$hash" | while read -r hash; do
+        sed -Ei.bak '/ *'"$1"' = "sha256-/s@".+"@"'"$hash"'"@' "$bin_file"
+        rm "$bin_file.bak"
+      done
+    done
+  done
```
Comment (thiagokokada) [inline pkgs/games/osu-lazer/update.sh:34]: Can we have a separate file for the `bin` version? Seems kinda messy to keep two different unrelated code in the same file.

File: pkgs/games/osu-lazer/bin.nix [hunk 2]
```diff
@@ -1,37 +1,84 @@
...
+    aarch64-darwin = "osu.app.Apple.Silicon.zip";
+    x86_64-darwin = "osu.app.Intel.zip";
+    x86_64-linux = "osu.AppImage";
+  }.${stdenv.system} or (throw "${pname}-${version}: ${stdenv.system} is unsupported.");
+
+  fileHash = {
+    aarch64-darwin = "sha256-vfDEEU+InD5uIc1QKceM2SN1GlWV93DySn3ik1MSYf4=";
+    x86_64-darwin = "sha256-dK+qqwBzuPXHqQk8U3TpWQBWEz+8S9GHShVwTTq2BBQ=";
+    x86_64-linux = "sha256-W3XJ7HtJM5iFI8OOTTu8IBHMerZSCETHMemkoTislK8=";
+  }.${stdenv.system};
```
Comment (thiagokokada) [inline pkgs/games/osu-lazer/bin.nix:23]: More of a personal preference, but I would have both information in the same attrset. Something like:

```suggestion
  file = {
    aarch64-darwin = {
      name = "osu.app.Apple.Silicon.zip";
      hash = "sha256-vfDEEU+InD5uIc1QKceM2SN1GlWV93DySn3ik1MSYf4="
    };
    x86_64-darwin = {
      name = "osu.app.Intel.zip";
      hash = "sha256-dK+qqwBzuPXHqQk8U3TpWQBWEz+8S9GHShVwTTq2BBQ=";
    };
    x86_64-linux = {
      name = "osu.AppImage";
      hash = "sha256-W3XJ7HtJM5iFI8OOTTu8IBHMerZSCETHMemkoTislK8=";
    };
  }.${stdenv.system} or (throw "${pname}-${version}: ${stdenv.system} is unsupported.");
```

**Comment by thiagokokada** [comment]
> About the cause: it seems like darwin systems are using fetchzip instead of fetchurl, the sha256 output is based on the unpacked file content instead of the zip file's sha256.

Try:

```
$ nix-prefetch-url --unpack https://github.com/ppy/osu/releases/download/2023.305.0/osu.app.Intel.zip
path is '/nix/store/8z4as2llcwzm9gzyrga1andynkw5q0gb-osu.app.Intel.zip'
0ir68bkc7m17yvqgl9s67s9bcpm1b7w23yk6d31jr5f9ih23rghj
```

I think there is a way to do the same with the new CLI, but I forgot.

**Comment by StepBroBD** [comment]
Thanks for the comment!

> nix-prefetch-url --unpack https://github.com/ppy/osu/releases/download/2023.305.0/osu.app.Intel.zip

This still outputs the incorrect sha256:

```
$ nix hash to-sri --type sha256 $(nix-prefetch-url --unpack https://github.com/ppy/osu/releases/download/2023.305.0/osu.app.Intel.zip)                                                                                                                                                                                                                                                                                                                                                                   
path is '/nix/store/473x8g03s583w0ai76c5666z799m6g3z-osu.app.Intel.zip'
sha256-7LV/Vr44ntNuNVBhmhMxdQjdX92+caCBULJ9jp/GJrk
```

```
$ nix-build -E "(import <nixpkgs> {}).fetchzip {url = \"https://github.com/ppy/osu/releases/download/2023.305.0/osu.app.Intel.zip\"; sha256=\"\"; stripRoot = false; }"                                                                                                                                                                                                                                                                                                                                  
......
unpacking source archive /private/tmp/nix-build-source.drv-0/osu.app.Intel.zip
error: hash mismatch in fixed-output derivation '/nix/store/x20jjkjqaxvaqlgralsj2rn3zh8bawxg-source.drv':
         specified: sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
            got:    sha256-Er48BIzJlSzDaGb6IfhZoV62kj5GJ/rw9ifUw+ZCJkc=
error: build of '/nix/store/rf0mkwvlvbqymjdlayfpzwb0kqc32ww1-source.drv' failed
```

**Comment by StepBroBD** [comment]
Duplicate of #222118, please close this PR.
